File: app/src/main/java/li/doerf/hacked/ui/adapters/AccountsAdapter.java
package li.doerf.hacked.ui.adapters;

import android.content.Context;
import android.content.Intent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.google.common.collect.Collections2;

import org.joda.time.format.DateTimeFormat;
import org.joda.time.format.DateTimeFormatter;

import java.util.Collection;
import java.util.List;

import androidx.annotation.NonNull;
import androidx.cardview.widget.CardView;
import androidx.fragment.app.FragmentManager;
import androidx.recyclerview.widget.RecyclerView;
import li.doerf.hacked.R;
import li.doerf.hacked.activities.BreachDetailsActivity;
import li.doerf.hacked.db.AppDatabase;
import li.doerf.hacked.db.daos.BreachDao;
import li.doerf.hacked.db.entities.Account;
import li.doerf.hacked.db.entities.Breach;
import li.doerf.hacked.ui.DeleteAccountDialogFragment;
import li.doerf.hacked.utils.BackgroundTaskHelper;
import li.doerf.hacked.utils.NotificationHelper;

public class AccountsAdapter extends RecyclerView.Adapter<RecyclerViewHolder> {
//    private final String LOGTAG = getClass().getSimpleName();
    private final FragmentManager mySupportFragmentManager;
    private final BreachDao myBreachDao;
    private List<Account> myAccountList;

    private final Context myContext;

    public AccountsAdapter(Context aContext, List<Account> accountList, FragmentManager supportFragmentManager){
        mySupportFragmentManager = supportFragmentManager;
        myAccountList = accountList;
        myContext = aContext;
        myBreachDao = AppDatabase.get(aContext).getBreachDao();
    }

    @Override
    public RecyclerViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        CardView itemLayout = (CardView) LayoutInflater.from(parent.getContext())
                .inflate(R.layout.card_account, parent, false);
        return new RecyclerViewHolder(itemLayout);
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerViewHolder holder, int position) {
        CardView cardView = (CardView) holder.getView();

        final Account account = myAccountList.get(position);

        new BackgroundTaskHelper<List<Breach>>().runInBackgroundAndConsumeOnMain(
                () -> myBreachDao.findByAccount(account.getId()),
                breaches -> bindView(cardView, account, breaches));
    }

    private void bindView(CardView cardView, Account account, Collection<Breach> breaches) {
        int numBreaches = breaches.size();
        int numAcknowledgedBreaches = Collections2.filter(breaches, Breach::getAcknowledged).size();
        TextView nameView = cardView.findViewById(R.id.name);
        nameView.setText(account.getName());


        TextView lastCheckedView = cardView.findViewById(R.id.last_checked);
        if ( account.getLastChecked() != null ) {
            DateTimeFormatter dtfOut = DateTimeFormat.forPattern("yyyy/MM/dd HH:mm");
            lastCheckedView.setText(dtfOut.print(account.getLastChecked()));
        } else {
            lastCheckedView.setText("-");
        }

        // set breach status text
        TextView breachStatus = cardView.findViewById(R.id.breach_state);
        if ( account.getLastChecked() == null ) {
            breachStatus.setText("-");
        } else if ( numBreaches == 0 ) {
            breachStatus.setText(getContext().getString(R.string.status_no_breach_found));
        } else {
            if ( numAcknowledgedBreaches == 0 ) {
                breachStatus.setText(getContext().getString(R.string.status_breaches_found, numBreaches));
            } else {
                breachStatus.setText(getContext().getString(R.string.status_breaches_found_acknowledged, numBreaches, numAcknowledgedBreaches));
            }
        }

        View statusIndicator = cardView.findViewById(R.id.status_indicator);
        // set color of card
        if ( account.getHacked()) {
            statusIndicator.setBackgroundColor(getContext().getResources().getColor(R.color.account_status_breached));
        } else if ( ! account.getHacked() && account.getLastChecked() == null ) {
            statusIndicator.setBackgroundColor(getContext().getResources().getColor(R.color.account_status_unknown));
        } else {
            if ( numBreaches == 0 ) {
                statusIndicator.setBackgroundColor(getContext().getResources().getColor(R.color.account_status_ok));
            } else {
                statusIndicator.setBackgroundColor(getContext().getResources().getColor(R.color.account_status_only_acknowledged));
            }
        }

        cardView.setOnClickListener(v -> {
            Intent showBreachDetails = new Intent(getContext(), BreachDetailsActivity.class);
            showBreachDetails.putExtra(BreachDetailsActivity.EXTRA_ACCOUNT_ID, account.getId());
            getContext().startActivity(showBreachDetails);
            NotificationHelper.cancelAll(getContext());
        });

        cardView.setOnLongClickListener(v -> {
            DeleteAccountDialogFragment d = new DeleteAccountDialogFragment();
            d.setAccount(account);
            d.show(mySupportFragmentManager, "deleteAccount");
            return true;
        });
    }

    @Override
    public int getItemCount() {
        return myAccountList.size();
    }

    public void addItems(List<Account> accountList) {
        myAccountList = accountList;
        notifyDataSetChanged();
    }

    private Context getContext() {
        return myContext;
    }
}

File: app/src/main/kotlin/li/doerf/hacked/ui/adapters/AccountsAdapter.kt
package li.doerf.hacked.ui.adapters

import android.content.Context
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import io.reactivex.processors.PublishProcessor
import li.doerf.hacked.HackedApplication
import li.doerf.hacked.R
import li.doerf.hacked.db.entities.Account
import li.doerf.hacked.util.NavEvent
import li.doerf.hacked.utils.NotificationHelper
import java.util.*


class AccountsAdapter(private val context: Context, private var accountList: List<Account>) : RecyclerView.Adapter<RecyclerViewHolder>() {

    private lateinit var navEvents: PublishProcessor<NavEvent>

    override fun onViewAttachedToWindow(holder: RecyclerViewHolder) {
        super.onViewAttachedToWindow(holder)
        navEvents = (context as HackedApplication).navEvents
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerViewHolder {
        val itemLayout = LayoutInflater.from(parent.context)
                .inflate(R.layout.card_account, parent, false)
        return RecyclerViewHolder(itemLayout)
    }

    override fun onBindViewHolder(holder: RecyclerViewHolder, position: Int) {
        val view = holder.view
        val account = accountList[position]
        bindView(view, account)
    }

    private fun bindView(view: View, account: Account) {
        val statusIndicator = view.findViewById<View>(R.id.status_indicator)
        setStatusIndicatorColor(account, statusIndicator)

        val nameView = view.findViewById<TextView>(R.id.name)
        nameView.text = account.name

        val breachCount = view.findViewById<TextView>(R.id.breach_count)
        val breachCounter = account.numBreaches
        Log.d(LOGTAG, "breachCounter: ${breachCounter}")

        if (breachCounter > 0) {
            breachCount.text = String.format(Locale.getDefault(), "%d", breachCounter)
            breachCount.visibility = View.VISIBLE
        } else {
            breachCount.visibility = View.GONE
        }

        view.setOnClickListener {
            NotificationHelper.cancelAll(context)
            navEvents.onNext(NavEvent(NavEvent.Destination.ACCOUNTS_DETAILS, account.id, null))
        }
    }

    private fun setStatusIndicatorColor(account: Account, statusIndicator: View) {
        if (account.hacked) {
            statusIndicator.setBackgroundColor(context.resources.getColor(R.color.account_status_breached))
        } else if (!account.hacked && account.lastChecked == null) {
            statusIndicator.setBackgroundColor(context.resources.getColor(R.color.account_status_unknown))
        } else {
            if (account.numBreaches == 0) {
                statusIndicator.setBackgroundColor(context.resources.getColor(R.color.account_status_ok))
            } else {
                statusIndicator.setBackgroundColor(context.resources.getColor(R.color.account_status_only_acknowledged))
            }
        }
    }

    override fun getItemCount(): Int {
        return accountList.size
    }

    fun addItems(accountList: List<Account>) {
        this.accountList = accountList
        notifyDataSetChanged()
    }

    companion object {
        private val LOGTAG = this::class.java.name
    }

}